### DISCUSSION TITLE
Instructors: Experiences from teaching the course

### DISCUSSION THREAD

**Discussion by mgeisler** [google/comprehensive-rust#86] [Show and tell]
Hi all!

I would love to hear from people outside of Google who use the course to bring colleagues up to speed on Rust. Please share your stories and experience here. It'll be valuable for the community to learn from this so we can improve the course for everybody.

**Comment by mgeisler** [comment]
See #156 for @CasperN's experience with the course.

**Comment by mo8it** [comment]
I published a blog post about teaching a modified version of Comprehensive Rust at my university:

https://mo8it.com/blog/teaching-rust/

It might be interesting to you @mgeisler 
Thanks for working on Comprehensive Rust 🥰

I actually want to contribute to upstream here as I did and want to further do in Rustlings. But the CLA and having to create a Google account for it is a big hurdle for me. Is there a way around it?

**Reply by mgeisler** [reply]
Wow, amazing! Thanks for sharing your experience! It seems you mostly liked the content, apart from the exercises — I very much agree, I think the exercises are the weakest part of the course right now. Using Rustlings is something we should look into!

> I actually want to contribute to upstream here as I did and want to further do in Rustlings. But the CLA and having to create a Google account for it is a big hurdle for me. Is there a way around it?

I dont't think so — the CLA page says that the Google account is used to authenticate you. It's our unified login system so I guess it makes sense that it's built on that.

I'll ask around since I would hate to see the course material becoming fragmented.

**Reply by mgeisler** [reply]
Hey @mo8it, I asked internally and the answer is that you must have a Google account to use the CLA tool, and we cannot merge PRs from you without the CLA... sorry about that.

You could create a Google account in ~5 minutes and [add your non-gmail.com address to the account](https://support.google.com/accounts/answer/176347). That's what I've done and the CLA robot is happy with that. So after creating the account, you can keep using your regular email for all your commits and you'll never have to touch the Google account again.

**Reply by mo8it** [reply]
Alright, thanks for asking ❤️

I will contribute when I find some time.

> I would hate to see the course material becoming fragmented.

I want to emphasize that my fork is just a temporary or soft fork. I am not trying to offer an alternative to this original :D

**Reply by mgeisler** [reply]
Thanks a lot, I am very excited to see others pick up the course. Your use of Rustlings is interesting and something we should look at more in #1082.

**Reply by azzamsa** [reply]
@mo8it In my initial course sessions, I originally relied on a blend of 'easy Rust' and 'comprehensive Rust.' However, upon reviewing your fork, I've chosen to use it as my primary source due to its material organization, which I find more appealing. https://github.com/azzamsa/blog/tree/main/src/talks/rust-course-bgn

**Reply by mgeisler** [reply]
@azzamsa, can I get you to work with @djmitche in #1073 to see if you can improve the ordering more? Also, note that we've been reordering the content ourselves (#913) recently.

I cannot prevent fragmentation, but I would really like the best ideas to flow back to this repository. So please submit PRs for anything we can fix and improve.

**Comment by mgeisler** [comment]
The course was mentioned on Mastodon: the course was used to [train ~20 engineers at Square/Block](https://elk.zone/hachyderm.io/@yerke/113314372295821447).

**Comment by mgeisler** [comment]
See #2502 for another report about teaching the course to about 15 Python developers.

**Comment by t-moe** [comment]
**Feedback after conducting a 4-day training (10 students, C#/C++ background)**

I recently used _Comprehensive Rust_ to facilitate an internal training session for a client team of 10 developers. The group primarily had C# and C++ backgrounds. We covered Days 1–4 plus the "Concurrency Day" (Threads + Async) over a 4-day period.

Overall, the feedback was very positive! However, based on my experience delivering the material as an instructor, I noticed a few areas where the curriculum could be further optimized:

* **Cargo & Project Structure**: The current introduction is quite lean. While `cargo new` works for local exercises, students would benefit from a broader overview of workspaces, packages vs. crates, and dependency management (including feature flags). Showing how to manage multiple binaries/examples in a single crate would also help them better organize their work.

* **Error Handling Timing**: I found that Error Handling (specifically `Result` and the `?` operator) comes a bit late (Day 4). Since many earlier topics touch on these concepts naturally, introducing them earlier would reduce friction during the first few days.

* **Exercise Alignment**: A few exercises felt slightly disconnected from the specific topic of the day. For example, the Protobuf exercise didn't seem to have a clear relation to the Lifetimes topic it was paired with.

**The "Developer/Manager" Perspective**: I started Day 1 by asking students about their preconceptions of Rust (from both a developer and manager perspective) and shared some success stories from the industry. I asked the same question again at the end, and the shift toward a positive outlook was very encouraging. This might be a useful "icebreaker" for other instructors to gauge the impact of the course.

Thanks to the maintainers for providing such a high-quality resource! I'll try to contribute improvements as time permits and as I re-teach this course.

**Reply by mgeisler** [reply]
Hi Timo, nice to meet you! Thanks a ton for writing up your feedback. Getting input from real life sessions is very helpful for us.

> * **Cargo & Project Structure**: The current introduction is quite lean.

I completely agree. What you see here is a combination of a few factors playing in when I first wrote this section:

- Android engineers don't use Cargo as part of their regular workflow, they use an Android build system called Soong.
- Google engineers also don't use Cargo, they use Bazel 😅
- I was hoping to not have to talk about which editor or local development environment to use since it seems like an open ended task... I use Emacs myself and figured that everybody would just use their own favorite setup. I think I underestimated this.

All this to say that I think a rewrite of the beginning of Day 1 would be in order.

I now teach Rust outside of Google, so I feel this pain too.

> * **Error Handling Timing**: I found that Error Handling (specifically `Result` and the `?` operator) comes a bit late (Day 4)

Yes, I see this too. I actually end up introducing it (briefly) earlier since people naturally ask about "exceptions" and error handling in general.

We could perhaps point out earlier when the instructor should mention these concepts?

> * **Exercise Alignment**

Yeah, I saw this too last time I taught the class. I ended up leaving out some exercises (the elevator one) due to lack of time and because I found it covered stuff that people probably already understood at this point.

**Reply by djmitche** [reply]
We have used a kind of "circular" design, where we touch on the same topic multiple times at different levels of depth. So, perhaps a lighter touch at error handling, earlier in the course, would make sense. Something like
 * C++/Java-style exceptions do not exist in Rust, and error conditions are handled explicitly
 * Errors are handled with a standard type called Result
 * You can panic on error with `.unwrap` or `.expect` or, in a function returning Result, use `?`
 * We will cover all of this in more depth after talking about generics.

Regarding the exercises -- this is absolutely true! Designing exercises is quite difficult and I certainly missed the mark on a number of them. PRs to introduce better exercises would be warmly welcomed :)

**Reply by mgeisler** [reply]
> Designing exercises is quite difficult and I certainly missed the mark on a number of them.

About the protobuf exercise: when I taught the class some weeks ago, some of the students loved it! They really enjoyed the challenge. But others were completely lost as it is non-trivial to say the least.

I would say that it's not just hard to make exercises, it's probably impossible to make exercises that match the skill levels of everybody.

Saying that, I get an idea: we could attempt to structure exercises so that they have a clear progression: a very easy step that lets the attendees type in a bit of Rust code related to the current chapter. It can be nearly trivial, it's meant as an easy win. Then a second step which asks more, and a third stretch goal which is really hard.

Doing this is of course in itself hard and requires real effort 😄 

Cc @gribozavr as well.

**Reply by t-moe** [reply]
Small clarification from my side: The students also loved the protobuf exercise (and all the other exercises). They only mentioned that the protobuf exercise had not much relation to the lifetime topic.
